internal/dataplane/translator/subtranslator/httproute.go

```diff
@@ -355,6 +603,224 @@ func mustMarshalJSON[T any](val T) string {
+	if err != nil {
+		return []kongstate.Route{}, err
+	}
+	if len(headers) > 0 {
+		r.Route.Headers = headers
+	}
+
+	// stripPath needs to be disabled by default to be conformant with the Gateway API
+	r.StripPath = kong.Bool(false)
+
+	// Check if the route has a RequestRedirect or URLRewrite with non-nil ReplacePrefixMatch - if it does, we need to
+	// generate a route for each match as the path is used to modify routes and generate plugins.
+	hasRedirectFilter := lo.ContainsBy(filters, func(filter gatewayapi.HTTPRouteFilter) bool {
+		return filter.Type == gatewayapi.HTTPRouteFilterRequestRedirect
+	})
+
+	routes, err := getRoutesFromMatches(matches, &r, filters, tags, hasRedirectFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	var path string
+	if hasURLRewriteWithReplacePrefixMatchFilter := lo.ContainsBy(filters, func(filter gatewayapi.HTTPRouteFilter) bool {
+		return filter.Type == gatewayapi.HTTPRouteFilterURLRewrite &&
+			filter.URLRewrite.Path != nil &&
+			filter.URLRewrite.Path.Type == gatewayapi.PrefixMatchHTTPPathModifier &&
+			filter.URLRewrite.Path.ReplacePrefixMatch != nil
+	}); hasURLRewriteWithReplacePrefixMatchFilter {
+		// In the case of URLRewrite with non-nil ReplacePrefixMatch, we rely on a CEL validation rule that disallows
+		// rules with multiple matches if the URLRewrite filter is present. We can be certain that if the filter is
+		// present, there is at most only one match. Based on that, we can determine the path from the first match.
+		// See: https://github.com/kubernetes-sigs/gateway-api/blob/29e68bffffb9af568e35545305d78d0001a1a0f7/apis/v1/httproute_types.go#L131
+		if len(matches) > 0 && matches[0].Path != nil && matches[0].Path.Value != nil {
+			path = *matches[0].Path.Value
+		}
+	}
+
+	// If the redirect filter has not been set, we still need to set the route plugins.
+	if !hasRedirectFilter {
+		if err := SetRoutePlugins(&r, filters, path, tags, false); err != nil {
+			return nil, err
+		}
+		routes = []kongstate.Route{r}
+	}
+
+	return routes, nil
+}
+
+func generateKongstateHTTPRoute(routeName string, ingressObjectInfo util.K8sObjectInfo, hostnames []*string) kongstate.Route {
+	// build the route object using the method and pathing information
+	r := kongstate.Route{
+		Ingress: ingressObjectInfo,
+		Route: kong.Route{
+			Name:         kong.String(routeName),
+			Protocols:    kong.StringSlice("http", "https"),
+			PreserveHost: kong.Bool(true),
+			// metadata tags aren't added here, they're added by the caller
+		},
+	}
+
+	// attach any hostnames associated with the httproute
+	if len(hostnames) > 0 {
+		r.Hosts = hostnames
+	}
+
+	return r
+}
+
+// convertGatewayMatchHeadersToKongRouteMatchHeaders takes an input list of Gateway APIs HTTPHeaderMatch
+// and converts these header matching rules to the format expected by go-kong.
+func convertGatewayMatchHeadersToKongRouteMatchHeaders(headers []gatewayapi.HTTPHeaderMatch) (map[string][]string, error) {
```

Comment:
I believe the intention was to move this function to `subtranslator` package, but now it's duplicated. Should we remove the one in `translator`?

Reply:
I had a commit to remove them in #6734 for implementing it in expression routes but did do the removal work in this PR in the later rounds of addressing comments. I would do a first round of removing in the later commits.